sdk/src/assertions/collection_hash.rs

```diff
@@ -0,0 +1,79 @@
+use std::io::{Read, Seek};
+
+use serde::{Deserialize, Serialize};
+
+use crate::{assertions::AssetType, asset_handlers::zip_io, hash_stream_by_alg, Error, Result};
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct CollectionHash {
+    pub uri_maps: Vec<UriHashedDataMap>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub alg: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none", with = "serde_bytes")]
+    pub zip_central_directory_hash: Option<Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct UriHashedDataMap {
+    pub uri: String,
+
+    #[serde(with = "serde_bytes")]
+    pub hash: Vec<u8>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub size: Option<u64>,
+
+    #[serde(rename = "dc:format", skip_serializing_if = "Option::is_none")]
+    pub dc_format: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub data_types: Option<Vec<AssetType>>,
+}
+
+impl CollectionHash {
```

Comment:
Collection hashes are not just for ZIPs.  This needs to be thought through on how to make this work in general.  I think there would need to be a new API to allow users to set the list of included external files or in the case of the zip something like this would happen automatically.

Reply:
I was thinking about that, I'll probably split the collection assertion off into a separate PR. Maybe we can allow the user to specify their own "URI resolver" of some such to fetch inclusion ranges. Then for ZIP, we can offer a predefined resolver.